flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/RowComparator.java

```diff
@@ -0,0 +1,698 @@
+import java.util.Collections;
+import java.util.List;
+
+import static org.apache.flink.api.java.typeutils.runtime.NullMaskUtils.readIntoNullMask;
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * Comparator for {@link Row}
+ */
+public class RowComparator extends CompositeTypeComparator<Row> {
+
+	private static final long serialVersionUID = 1L;
+	/** The number of fields of the Row */
+	private final int arity;
+	/** key positions describe which fields are keys in what order */
+	private final int[] keyPositions;
+	/** null-aware comparators for the key fields, in the same order as the key fields */
+	private final NullAwareComparator<Object>[] comparators;
+	/** serializers to deserialize the first n fields for comparison */
+	private final TypeSerializer<Object>[] serializers;
+	/** auxiliary fields for normalized key support */
+	private final int[] normalizedKeyLengths;
+	private final int numLeadingNormalizableKeys;
+	private final int normalizableKeyPrefixLen;
+	private final boolean invertNormKey;
+
+	// null masks for serialized comparison
+	private final boolean[] nullMask1;
+	private final boolean[] nullMask2;
+
+	// cache for the deserialized key field objects
+	transient private final Object[] deserializedKeyFields1;
+	transient private final Object[] deserializedKeyFields2;
+
+	/**
+	 * General constructor for RowComparator.
+	 *
+	 * @param arity        the number of fields of the Row
+	 * @param keyPositions key positions describe which fields are keys in what order
+	 * @param comparators  non-null-aware comparators for the key fields, in the same order as
+	 *                     the key fields
+	 * @param serializers  serializers to deserialize the first n fields for comparison
+	 * @param orders       sorting orders for the fields
+	 */
+	public RowComparator(
+		int arity,
+		int[] keyPositions,
+		TypeComparator<Object>[] comparators,
+		TypeSerializer<Object>[] serializers,
+		boolean[] orders) {
+		this(arity, keyPositions, makeNullAware(comparators, orders), serializers);
+	}
+
+
+	/**
+	 * Intermediate constructor for creating auxiliary fields.
+	 */
+	private RowComparator(
+		int arity,
+		int[] keyPositions,
+		NullAwareComparator<Object>[] comparators,
+		TypeSerializer<Object>[] serializers) {
+		this(
+			arity,
+			keyPositions,
+			comparators,
+			serializers,
+			createAuxiliaryFields(keyPositions, comparators));
+	}
+
+	/**
+	 * Intermediate constructor for creating auxiliary fields.
+	 */
+	private RowComparator(
+		int arity,
+		int[] keyPositions,
+		NullAwareComparator<Object>[] comparators,
+		TypeSerializer<Object>[] serializers,
+		Tuple4<int[], Integer, Integer, Boolean> auxiliaryFields) {
+		this(
+			arity,
+			keyPositions,
+			comparators,
+			serializers,
+			auxiliaryFields.f0,
+			auxiliaryFields.f1,
+			auxiliaryFields.f2,
+			auxiliaryFields.f3);
+	}
+
+	/**
+	 * Intermediate constructor for creating auxiliary fields.
+	 */
+	private RowComparator(
```

Comment:
This constructor is used in `duplicate` which I don't want to generate the Tuple4 again.